src/storage-client/src/controller.rs

```diff
@@ -3168,44 +3170,132 @@ where
-            // TODO(aljoscha): This comes largely from the fact that
-            // subsources are created with a `DataSource::Other`, so
-            // we have no idea (at their creation time) that they
-            // are a subsource, or that they are a subsource of a
-            // source where they need a read hold on that
-            // ingestion's remap collection.
+            // TODO(later): remove this if statement once we fix the inverse
+            // dependency of subsources
             if PartialOrder::less_than(&collection.implied_capability, &dependency_since) {
+                // TODO(next release): assert!(*collection.write_frontier ==
+                // [T::minimum()])
                 assert!(
-                    PartialOrder::less_than(&dependency_since, &collection.write_frontier),
-                    "write frontier ({:?}) must be in advance dependency collection's since ({:?})",
+                    // This assertion is meant to mirror
+                    // https://github.com/MaterializeInc/materialize/blob/b997fa24db04bfaf4465421e814ceff167543fa0/src/storage/src/source/reclock/compat.rs#L117,
+                    PartialOrder::less_than(&dependency_since, &collection.write_frontier)
+                        // Accepting minimum values lets us advance new
+                        // subsource's sinces to the dependency since--this
+                        // collection's write frontier being the minimum will
+                        // ultimately not affect the collection's as_of, so this
+                        // is fine to allow.
+                        || *collection.write_frontier == [T::minimum()],
+                    "write frontier ({:?}) must be @ min or in advance of dependency collection's since ({:?})",
                     collection.write_frontier,
                     dependency_since,
                 );
+
+                // TODO(next release): assert!(*collection.implied_capbility ==
+                // [T::minimum()])
                 mz_ore::soft_assert!(
-                    matches!(collection.read_policy, ReadPolicy::NoPolicy { .. }),
+                    match &collection.read_policy {
+                        ReadPolicy::NoPolicy { initial_since } =>
+                        // TODO(next release): assert!(*initial_since ==
+                        // [T::minimum()])
+                            PartialOrder::less_than(initial_since, &dependency_since),
+                        _ => false,
+                    },
                     "subsources should not have external read holds installed until \
                                     their ingestion is created, but {:?} has read policy {:?}",
                     id,
                     collection.read_policy
                 );
 
-                // This patches up the implied_capability!
+                // Downgrade the since
+                let mut read_handle: ReadHandle<SourceData, (), T, Diff> = persist_client
```

Comment:
I think we have to use a (critical) `SinceHandle` with the same ID that the storage/handle worker thing uses to make sure that the downgrade actually goes through. And it might be worth adding a soft assert that ensures that the overall shard since really is what we expect here.

Reply:
great catch, ty